### PR TITLE
Manually check HDF5 files

### PR DESCRIPTION
**Pull request by graeme-winter**
Look for typically useless but easily passed in files such as the data block, meta file. Fixes dials/dials#1220

### CODE REVIEW

File: format/FormatHDF5.py [hunk 1]
```diff
@@ -18,7 +19,12 @@ def __init__(self, image_file, **kwargs):
...
         try:
             with FormatHDF5.open_file(image_file, "rb") as fh:
-                return fh.read(8) == b"\211HDF\r\n\032\n"
+                if fh.read(8) != b"\211HDF\r\n\032\n":
+                    return False
+            if hdf5_file_type(image_file) == HDF5_NXS_FILE:
+                return True
+            else:
+                return False
         except IOError:
```
Comment (Anthchirp) [inline format/FormatHDF5.py:28]: If we try to `h5py.File`-open then we don't need to do a separate signature check, as the HDF5 library will do that anyway.
```python
try:
    return get_hdf5_file_type(image_file) is HDF5_FileType.NXS
except (IOError, OSError):
    return False
```

File: format/FormatHDF5Exceptions.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import h5py
+
+# static enumerations for useless file types
+
+HDF5_UNKNOWN = -1
+HDF5_NXS_FILE = 0
+HDF5_DATA_FILE = 1
+HDF5_META_FILE = 2
+HDF5_MPCCD = 0
```
Comment (dagewa) [inline format/FormatHDF5Exceptions.py:9]: I wonder if encapsulation using the [enum](https://docs.python.org/3/library/enum.html) library might be preferable?

File: format/FormatHDF5Exceptions.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+        # NeXus type files contain only /entry at the top (at the moment)
+        if top_level == set(["entry"]):
+            return HDF5_NXS_FILE
+
+        # SACLA MPCCD contains "metadata" and ... other stuff
+        if "metadata" in top_level:
+            return HDF5_MPCCD
+
+    return HDF5_UNKNOWN
```
Comment (Anthchirp) [inline format/FormatHDF5Exceptions.py:34]:
```suggestion
from enum import Enum, auto
class HDF5_FileType(Enum):
    DATA=auto()
    META=auto()
    MPCCD=auto()
    NXS=auto()
    UNKNOWN=auto()


def get_hdf5_file_type(hdf5_filename):
    """Open the named file, decide if it is probably not a useful one for
    processing..."""

    with h5py.File(hdf5_filename, "r") as f:
        top_level = set(f.keys())
        # case HDF5 data file
        if top_level == {"data"}:
            return HDF5_FileType.DATA

        # case Diamond "meta" file
        if "config" in top_level and "frame_written" in top_level:
            return HDF5_FileType.META

        # NeXus type files contain only /entry at the top (at the moment)
        if top_level == {"entry"}:
            return HDF5_FileType.NXS

        # SACLA MPCCD contains "metadata" and ... other stuff
        if "metadata" in top_level:
            return HDF5_FileType.MPCCD

    return HDF5_FileType.UNKNOWN
```

File: format/FormatHDF5Exceptions.py (hunk 2 above)
Comment (graeme-winter) [reply format/FormatHDF5Exceptions.py:9]: 👍 

TIL this

File: format/FormatHDF5.py [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 from dxtbx import IncorrectFormatError
 from dxtbx.format.Format import Format
+from dxtbx.format.FormatHDF5Exceptions import HDF5_NXS_FILE, hdf5_file_type
```
Comment (Anthchirp) [inline format/FormatHDF5.py:7]:
```suggestion
from dxtbx.format.FormatHDF5Exceptions import HDF5_FileType, get_hdf5_file_type
```

**Review by Anthchirp — APPROVED** [review]
* avoid double signature check
* enums are a thing

File: format/FormatHDF5Exceptions.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+# static enumerations for useless file types
+
+from enum import Enum, auto
+class HDF5_FileType(Enum):
+    DATA=auto()
+    META=auto()
+    MPCCD=auto()
+    NXS=auto()
+    UNKNOWN=auto()
```
Comment (Anthchirp) [inline format/FormatHDF5Exceptions.py:11]: I just found out that `Enum` allows `namedtuple`-style invocation which is much cleaner:
```suggestion
from enum import Enum
HDF5_FileType = Enum("HDF5_FileType", "DATA META MPCCD NXS UNKNOWN")
```

**Comment by graeme-winter** [comment]
@Anthchirp  thanks for your additions, have worked those in

**Comment by phyy-nx** [comment]
Not a fan of this PR, TBH.  Aside from breaking FormatHDF5PAL, it short-circuits the entire `understand` API by asserting control of what can be an h5 file at the FormatHDF5 level.  It moves `understand` in the wrong direction, making FormatHDF5 enforce what can be understood by sub-classes.

There's got to be a better way of fixing dials/dials#1220.

**Comment by graeme-winter** [comment]
> Not a fan of this PR, TBH. Aside from breaking FormatHDF5PAL, it short-circuits the entire `understand` API by asserting control of what can be an h5 file at the FormatHDF5 level. It moves `understand` in the wrong direction, making FormatHDF5 enforce what can be understood by sub-classes.
> 
> There's got to be a better way of fixing [dials/dials#1220](https://github.com/dials/dials/issues/1220).

@phyy-nx am pretty respectful of your viewpoint here - it felt bad doing this but at the same time it is *a* solution to the problem even if it is is not a pretty one

I'm pretty happy that people explore other avenues, and happy to close this PR if a better solution appears.

**Comment by ndevenish** [comment]
Superseded by #255 and #261

**Pull request closed without merging**